src/include/duckdb/common/types/string.hpp

```diff
@@ -0,0 +1,334 @@
+//===----------------------------------------------------------------------===//
+//                         DuckDB
+//
+// duckdb/common/types/string.hpp
+//
+//
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include <string>
+#include <cstring>
+
+#include "duckdb/common/numeric_utils.hpp"
+#include "duckdb/common/typedefs.hpp"
+#include "duckdb/common/assert.hpp"
+
+namespace duckdb {
+
+class String {
+	/*
+	    String either can own its data (with automatic cleanup) or hold a non-owning reference.
+	    * Move only semantics, no copying allowed.
+	    * Small Strings (≤7 bytes) are stored inline to avoid heap allocation.
+	    * 32-bit length limit with overflow protection.
+
+	    To create an owning String, use the constructors.
+	    To create a non-owning String, use the Reference methods.
+	 */
+
+public:
+	// Constructors (create an owning String)
+	String() : len(0), buf {0} {
+	}
+
+	String(const std::string &str) { // NOLINT allowimplicit conversion
+		AssignOwning(str.data(), SafeStrlen(str));
+	}
+
+	String(const char *str, const uint32_t size) {
+		if (str == nullptr) {
+			AssignOwning(str, 0);
+			return;
```

Comment:
This seems a bit superfluous, it's going to hit the assertion in AssignOwning, but if you really want to make it safe I would change it to do handle this case in AssignOwning instead (and remove the assertion)